wsmaster/che-core-api-workspace/src/main/java/org/eclipse/che/api/workspace/server/WorkspaceRuntimes.java

```diff
@@ -272,23 +256,74 @@ public WorkspaceStatus getStatus(String workspaceId) {
     }
   }
 
+  private class StartRuntimeTask implements Runnable {
+    private final Workspace workspace;
+    private final Map<String, String> options;
+    private final String workspaceId;
+    private final InternalRuntime runtime;
+
+    public StartRuntimeTask(
+        Workspace workspace, Map<String, String> options, InternalRuntime runtime) {
+      this.workspace = workspace;
+      this.options = options;
+      this.workspaceId = workspace.getId();
+      this.runtime = runtime;
+    }
+
+    @Override
+    public void run() {
+      try {
+        runtime.start(options);
+        runtimes.replace(workspaceId, new RuntimeState(runtime, RUNNING));
+
+        LOG.info(
+            "Workspace '{}:{}' with id '{}' started by user '{}'",
+            workspace.getNamespace(),
+            workspace.getConfig().getName(),
+            workspace.getId(),
+            sessionUserNameOr("undefined"));
+        publishWorkspaceStatusEvent(workspaceId, RUNNING, STARTING, null);
+      } catch (InfrastructureException e) {
+        runtimes.remove(workspaceId);
+        String failureCause = "failed";
+        if (e instanceof RuntimeStartInterruptedException) {
+          failureCause = "interrupted";
+          publishWorkspaceStatusEvent(workspaceId, STOPPED, STARTING, e.getMessage());
+        }
+        LOG.info(
+            "Workspace '{}:{}' with id '{}' start {}",
+            workspace.getNamespace(),
+            workspace.getConfig().getName(),
+            workspace.getId(),
+            failureCause);
+        // InfrastructureException is supposed to be an exception that can't be solved
+        // by Che admin, so should not be logged (but not InternalInfrastructureException).
+        // It will prevent bothering the admin when user made a mistake in WS configuration.
+        if (e instanceof InternalInfrastructureException) {
+          LOG.error(e.getLocalizedMessage(), e);
+        }
+        throw new RuntimeException(e);
```

Comment:
Is there any reason to throw the exception here? Maybe logging is enough?

Reply:
It is needed to propagate the result (failure or success) to a possible calling side that may be interested in chaining other stages depending on success/failure of this task.